### PR TITLE
browser(firefox): implement screencast

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: browser_patches/firefox/juggler/screencast/nsScreencastService.cpp [hunk 1]
```diff
@@ -101,26 +123,101 @@ class nsScreencastService::Session : public rtc::VideoSinkInterface<webrtc::Vide
     }
 
     mCaptureModule->RegisterCaptureDataCallback(this);
```
Comment (yury-s) [inline browser_patches/firefox/juggler/screencast/nsScreencastService.cpp:125]:
```
if (mEncoder)
  mCaptureModule->RegisterCaptureDataCallback(this);
else
  mCaptureModule->RegisterRawFrameCallback(this);
```

File: browser_patches/firefox/juggler/screencast/nsScreencastService.cpp [hunk 2]
```diff
@@ -80,11 +99,14 @@ nsresult generateUid(nsString& uid) {
...
 
-class nsScreencastService::Session : public rtc::VideoSinkInterface<webrtc::VideoFrame> {
+class nsScreencastService::Session : public rtc::VideoSinkInterface<webrtc::VideoFrame>,
+                                     public webrtc::RawFrameCallback {
  public:
-  Session(rtc::scoped_refptr<webrtc::VideoCaptureModule>&& capturer, RefPtr<ScreencastEncoder>&& encoder)
-      : mCaptureModule(std::move(capturer))
-      , mEncoder(std::move(encoder)) {
+  Session(const nsAString& sessionId, rtc::scoped_refptr<webrtc::VideoCaptureModuleEx>&& capturer, RefPtr<ScreencastEncoder>&& encoder, uint32_t jpegQuality)
+      : mSessionId(sessionId)
```
Comment (yury-s) [inline browser_patches/firefox/juggler/screencast/nsScreencastService.cpp:106]: mSessionId is unused, drop it

File: browser_patches/firefox/juggler/TargetRegistry.js [hunk 3]
```diff
@@ -502,38 +503,68 @@ class PageTarget {
...
+    const docShell = this._gBrowser.ownerGlobal.docShell;
+    // Exclude address bar and navigation control from the video.
+    const rect = this.linkedBrowser().getBoundingClientRect();
+    const devicePixelRatio = this._window.devicePixelRatio;
+    const screencastId = screencast.startVideoRecording(docShell, false, '', width, height, quality || 90, devicePixelRatio * rect.top);
+    const onFrame = (subject, topic, data) => {
+      this.emit(PageTarget.Events.ScreencastFrame, data);
+    };
+    Services.obs.addObserver(onFrame, 'juggler-screencast-frame');
+    this._screencastRecordingInfo = { screencast, screencastId, onFrame };
```
Comment (yury-s) [inline browser_patches/firefox/juggler/TargetRegistry.js:542]: Just initialize `screencast` at the top of the file and use the variable in the methods?

**Review by yury-s — APPROVED** [review]

File: browser_patches/firefox/juggler/screencast/nsScreencastService.cpp (hunk 1 above)
Comment (pavelfeldman) [reply browser_patches/firefox/juggler/screencast/nsScreencastService.cpp:126]: Done

File: browser_patches/firefox/juggler/screencast/nsScreencastService.cpp (hunk 2 above)
Comment (pavelfeldman) [reply browser_patches/firefox/juggler/screencast/nsScreencastService.cpp:106]: Done

File: browser_patches/firefox/juggler/TargetRegistry.js (hunk 3 above)
Comment (pavelfeldman) [reply browser_patches/firefox/juggler/TargetRegistry.js:542]: Done

**Pull request merged**